src/realm/object-store/sync/sync_session.cpp

```diff
@@ -402,7 +402,19 @@ SyncSession::SyncSession(Private, SyncClient& client, std::shared_ptr<DB> db, co
 
 void SyncSession::detach_from_sync_manager()
 {
-    shutdown_and_wait();
+    // Unregister all callbacks when the App and SyncManager are destroyed.
+    {
+        util::CheckedLockGuard lk(m_state_mutex);
+        m_completion_callbacks.clear();
+    }
+    {
+        util::CheckedLockGuard lk(m_connection_state_mutex);
+        m_connection_change_notifier.remove_callbacks();
+    }
+    m_progress_notifier.unregister_callbacks();
+
+    static constexpr bool cancel_subscription_notifications = false;
+    shutdown_and_wait(cancel_subscription_notifications);
```

Comment:
Shouldn't this be sending OperationCancelled to any subscription change notifications?

Reply:
my reasoning was to guard against `sub.get_state_change_notification().get_async()` use cases.

Reply:
Yes, that needs to invoke the completion handler. Every Future must always be fulfilled.

Reply:
but the completion handler may be gone and crash (pretty much the issue with every handler/callback).

Reply:
I see. I will revert this specific change then cc @gagik